From 0b7df7ba96c72dfb5c72727c3acb16c73418381e Mon Sep 17 00:00:00 2001
From: Olliver Schinagl <oliver@schinagl.nl>
Date: Fri, 24 Mar 2023 15:36:24 +0100
Subject: [PATCH] mtd: spi-nor: Add Fudan fm25q128a/fm25q64

This patch adds support for the Fudan Microelectronics FM25Q128A (128
Mbit, 16 MiByte) and FM25Q64 (64 Mbit, 8 MiByte) serial flash memory with
100MHz fast (DDR) read, 50MHz max read frequency.

It supports both single, dual and quad SPI interfaces and has a wide IO
range.

Tested on the Hasvio S1100 ethernet Switch.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>
---
 drivers/mtd/spi-nor/Makefile |  1 +
 drivers/mtd/spi-nor/core.c   |  1 +
 drivers/mtd/spi-nor/core.h   |  1 +
 drivers/mtd/spi-nor/fudan.c  | 30 ++++++++++++++++++++++++++++++
 4 files changed, 33 insertions(+)
 create mode 100644 drivers/mtd/spi-nor/fudan.c

diff --git a/drivers/mtd/spi-nor/Makefile b/drivers/mtd/spi-nor/Makefile
index e344077e3054..35f350fa84f7 100644
--- a/drivers/mtd/spi-nor/Makefile
+++ b/drivers/mtd/spi-nor/Makefile
@@ -6,6 +6,7 @@ spi-nor-objs			+= catalyst.o
 spi-nor-objs			+= eon.o
 spi-nor-objs			+= esmt.o
 spi-nor-objs			+= everspin.o
+spi-nor-objs			+= fudan.o
 spi-nor-objs			+= fujitsu.o
 spi-nor-objs			+= gigadevice.o
 spi-nor-objs			+= intel.o
diff --git a/drivers/mtd/spi-nor/core.c b/drivers/mtd/spi-nor/core.c
index 3ac1cc4af8dd..f57686fca59c 100644
--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -1847,6 +1847,7 @@ static const struct spi_nor_manufacturer *manufacturers[] = {
 	&spi_nor_eon,
 	&spi_nor_esmt,
 	&spi_nor_everspin,
+	&spi_nor_fudan,
 	&spi_nor_fujitsu,
 	&spi_nor_gigadevice,
 	&spi_nor_intel,
diff --git a/drivers/mtd/spi-nor/core.h b/drivers/mtd/spi-nor/core.h
index fd2771fbe710..e12e009c79be 100644
--- a/drivers/mtd/spi-nor/core.h
+++ b/drivers/mtd/spi-nor/core.h
@@ -477,6 +477,7 @@ extern const struct spi_nor_manufacturer spi_nor_catalyst;
 extern const struct spi_nor_manufacturer spi_nor_eon;
 extern const struct spi_nor_manufacturer spi_nor_esmt;
 extern const struct spi_nor_manufacturer spi_nor_everspin;
+extern const struct spi_nor_manufacturer spi_nor_fudan;
 extern const struct spi_nor_manufacturer spi_nor_fujitsu;
 extern const struct spi_nor_manufacturer spi_nor_gigadevice;
 extern const struct spi_nor_manufacturer spi_nor_intel;
diff --git a/drivers/mtd/spi-nor/fudan.c b/drivers/mtd/spi-nor/fudan.c
new file mode 100644
index 000000000000..5eb7661c1ff8
--- /dev/null
+++ b/drivers/mtd/spi-nor/fudan.c
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (C) 2023, Olliver Schinagl <oliver@schinagl.nl>
+ */
+
+#include <linux/mtd/spi-nor.h>
+
+#include "core.h"
+
+static const struct flash_info fudan_parts[] = {
+	/* Shanghai Fudan Microlectronics */
+	{ "fm25q64",    INFO(0xa14017, 0, 64 * 1024,   128,
+	                     SECT_4K |
+	                     SPI_NOR_DUAL_READ |
+	                     SPI_NOR_QUAD_READ |
+	                     SPI_NOR_HAS_LOCK |
+	                     SPI_NOR_HAS_TB) },
+	{ "fm25q128a",  INFO(0xa14018, 0, 64 * 1024,   256,
+	                     SECT_4K |
+	                     SPI_NOR_DUAL_READ |
+	                     SPI_NOR_QUAD_READ |
+	                     SPI_NOR_HAS_LOCK |
+	                     SPI_NOR_HAS_TB) },
+};
+
+const struct spi_nor_manufacturer spi_nor_fudan = {
+	.name = "fudan",
+	.parts = fudan_parts,
+	.nparts = ARRAY_SIZE(fudan_parts),
+};
-- 
2.39.2

